pkg/api/soc.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2021 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package api
+
+import (
+	"encoding/hex"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/ethersphere/bee/pkg/bmtpool"
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/soc"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/gorilla/mux"
+)
+
+type socPostResponse struct {
+	Reference swarm.Address `json:"reference"`
+}
+
+func (s *server) socSigUploadHandler(w http.ResponseWriter, r *http.Request) {
+	owner, err := hex.DecodeString(mux.Vars(r)["owner"])
+	if err != nil {
+		s.Logger.Debugf("soc upload: bad owner: %v", err)
+		s.Logger.Error("soc upload: bad owner")
+		jsonhttp.BadRequest(w, "bad owner")
+		return
+	}
+	id, err := hex.DecodeString(mux.Vars(r)["id"])
+	if err != nil {
+		s.Logger.Debugf("soc upload: bad id: %v", err)
+		s.Logger.Error("soc upload: bad id")
+		jsonhttp.BadRequest(w, "bad id")
+		return
+	}
+	sig, err := hex.DecodeString(mux.Vars(r)["sig"])
+	if err != nil {
+		s.Logger.Debugf("soc upload: bad signature: %v", err)
+		s.Logger.Error("soc upload: bad signature")
+		jsonhttp.BadRequest(w, "bad signature")
+		return
+	}
+
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		if jsonhttp.HandleBodyReadError(err, w) {
+			return
+		}
+		s.Logger.Debugf("soc upload: read chunk data error: %v", err)
+		s.Logger.Error("soc upload: read chunk data error")
+		jsonhttp.InternalServerError(w, "cannot read chunk data")
+		return
+	}
+	if len(data) < swarm.SpanSize {
+		s.Logger.Debugf("soc upload: chunk data too short")
+		s.Logger.Error("soc upload: chunk data")
+		jsonhttp.BadRequest(w, "short chunk data")
```

Comment:
https://gateway.ethswarm.org/bzz/latest.bookofswarm.eth/#subsection.10.2.1

201   CreatedReference in response body
400  Bad request     owner, id or span is not well formed
401   Unauthorized signing fails
404  Not found  no owner keypair is found
413  Payload Too Large Payload size exceeds span value or 4096
422  Unprocessable entity    owner/id pair already exists